govcd/user.go

```diff
@@ -0,0 +1,463 @@
+// (This operation does not return a task)
+// Note that the timeout is not absolute, meaning that we only wait the
+// full amount of the timeout if the user has not been created until then.
+// This function returns as soon as the user has been created, which could be as
+// little as 50ms or as much as TimeOut - 50 ms
+// Mandatory fields are: Name, Role, Password.
+// https://code.vmware.com/apis/442/vcloud-director#/doc/doc/operations/POST-CreateUser.html
+func (adminOrg *AdminOrg) CreateUser(userConfiguration *types.User, timeOut time.Duration) (*OrgUser, error) {
+	err := validateUserForCreation(userConfiguration)
+	if err != nil {
+		return nil, err
+	}
+
+	userCreateHREF, err := url.ParseRequestURI(adminOrg.AdminOrg.HREF)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing admin org url: %s", err)
+	}
+	userCreateHREF.Path += "/users"
+
+	user := NewUser(adminOrg.client, adminOrg)
+
+	_, err = adminOrg.client.ExecuteRequest(userCreateHREF.String(), http.MethodPost,
+		types.MimeAdminUser, "error creating user: %s", userConfiguration, user.User)
+	if err != nil {
+		return nil, err
+	}
+
+	// Uncomment this line to show the user structure after creation
+	// ShowUser(*user.OrgUser)
+
+	// If there is a valid task, we try to follow through
+	if user.User.Tasks != nil && len(user.User.Tasks.Task) > 0 {
+		task := NewTask(adminOrg.client)
+		task.Task = user.User.Tasks.Task[0]
+		err = task.WaitTaskCompletion()
+
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Attempting to retrieve the user
+	// Since there is no task to wait for, we try to retrieve using a max of 200 attempts
+	// at intervals of 50 Ms
+	// Experience shows that it usually takes less than one second to get the result
+	var delayPerAttempt time.Duration = 50
+	maxAttempts := 200
+	if timeOut > 0 && timeOut > delayPerAttempt {
+		maxAttempts = int(timeOut / delayPerAttempt)
+	}
+
+	// We make sure that the timeout is never less than 1 second
+	if maxAttempts < 20 {
+		maxAttempts = 20
+	}
+
+	startTime := time.Now()
+	var newUser *OrgUser
+	for N := 0; N <= int(maxAttempts); N++ {
+		newUser, err = adminOrg.GetUserByNameOrId(userConfiguration.Name, true)
+		if err == nil {
+			break
+		}
+		time.Sleep(delayPerAttempt)
+	}
+	endTime := time.Now()
+
+	elapsed := endTime.Sub(startTime)
+
+	// Uncomment this line to show how long the creation takes
+	// fmt.Printf("# %s: %s\n",user.OrgUser.Name,elapsed)
+
+	// If the user was not retrieved within the allocated time, we inform the user about the failure
+	// and the time it occurred to get to this point, so that they may try with a longer time
+	if err != nil {
+		return nil, fmt.Errorf("failure to retrieve a new user after %s : %s", elapsed, err)
+	}
+
+	return newUser, nil
+}
+
+// SimpleCreateUser creates an org user from a simplified structure
+func (adminOrg *AdminOrg) SimpleCreateUser(userData OrgUserConfiguration) (*OrgUser, error) {
+
+	if userData.Name == "" {
```

Comment:
Not fully sure, but maybe this could rely on underlying function `validateUserForCreation` which is called during the wrapped call `CreateUser`. That way all `create` validation would be in one place.

Reply:
`validateUserForCreation` uses a different structure. Here we're dealing with the simplified one.